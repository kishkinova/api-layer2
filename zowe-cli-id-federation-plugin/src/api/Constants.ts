/**
 * This program and the accompanying materials are made available and may be used, at your option, under either:
 * * Eclipse Public License v2.0, available at https://www.eclipse.org/legal/epl-v20.html, OR
 * * Apache License, version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Copyright Contributors to the Zowe Project.
 */

export class Constants {

    // Parameters maximum length
    //TODO: length valid for RACF, if other ESMs allow the different length -> refactor
    public static readonly maxLengthRegistry = 255;
    public static readonly maxLengthSystem = 8;

    // Plugin return codes
    public static readonly okayCode = 0;
    public static readonly zoweErrorCode = 1;
    public static readonly warnCode = 4;
    public static readonly fatalCode = 16;

    // CSV
    public static readonly headers = ['userName', 'distributedId', 'mainframeId'];

}
