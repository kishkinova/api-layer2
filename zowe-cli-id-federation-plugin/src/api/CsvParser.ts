/**
 * This program and the accompanying materials are made available and may be used, at your option, under either:
 * * Eclipse Public License v2.0, available at https://www.eclipse.org/legal/epl-v20.html, OR
 * * Apache License, version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Copyright Contributors to the Zowe Project.
 */

import * as fs from "fs";
import {parse} from "csv-parse/sync";
import {ImperativeError} from "@zowe/imperative";

export interface IIdentity {
    userName: string;
    distributedId: string;
    mainframeId: string;
}

export class CsvParser {

    constructor(
        public file: string) {
    }

    getIdentities(): IIdentity[] {
        let fileContent;
        try {
            fileContent = fs.readFileSync(this.file);
        } catch (error) {
            throw new ImperativeError({msg: `Cannot open the input CSV file. ${error.message}`});
        }
        const headers = ['userName', 'distributedId', 'mainframeId'];

        try {
            return parse(fileContent, {columns: headers}) as IIdentity[];
        } catch (e) {
            throw new ImperativeError({msg: `Invalid CSV format: ${e.message ?? ''}`});
        }
    }

}
