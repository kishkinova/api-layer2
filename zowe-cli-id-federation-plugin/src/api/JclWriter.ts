/**
 * This program and the accompanying materials are made available and may be used, at your option, under either:
 * * Eclipse Public License v2.0, available at https://www.eclipse.org/legal/epl-v20.html, OR
 * * Apache License, version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Copyright Contributors to the Zowe Project.
 */

const MAX_LINE_LENGTH = 71;

export class JclWriter {

    private commands: string = '';

    constructor(
        private firstIdent: number = 4,
        private nextIdent: number = 2
    ) {
    }

    parse(command: string): string[] {
        const words: string[] = [];
        let lastWord = "";
        const status = 0;
        lastWord = JclWriter.getLastWord(command, status, lastWord, words);
        // if lastWord still contains a word add it to response (text without spaces on end or even opened string)
        if (lastWord) words.push(lastWord);
        return words;
    }

    private static getLastWord(command: string, status: number, lastWord: string, words: string[]) {
        let i = 0;
        while (i < command.length) {
            if (status === 0) {// outside the string
                [status, lastWord] = JclWriter.parseCommand(command, i, status, lastWord, words);
            } else if (status === 1) {// inside the string
                lastWord += command.charAt(i);
                if (command.charAt(i) == '\'') {
                    if ((command.length > i + 1) && (command.charAt(i + 1) == '\'')) {
                        // two apos means escaped apos, still parsing the string
                        lastWord += '\'';
                        i++;
                    } else {
                        // ending apos, stop parsing the string
                        status = 0;
                    }
                }
            }
            i++;
        }
        return lastWord;
    }

    private static parseCommand(command: string, i: number, status: number, lastWord: string, words: string[]) {
        if (command.charAt(i) == '\'') {
            // on apos switch to state 1 - parsing a string
            status = 1;
            lastWord += '\'';
        } else if (command.charAt(i) == ' ') {
            // space means to store lastWord as fully parsed and wait for the next word
            if (lastWord) {
                words.push(lastWord);
                lastWord = '';
            }
        } else {
            // everything else is a part of the word
            lastWord += command.charAt(i);
        }
        return [status, lastWord] as const;
    }

    format(words: string[]): string {
        let formatted = "".padStart(this.firstIdent, ' ');
        let position = this.firstIdent;
        let blank = true;

        for (let i = 0; i < words.length; i++) {
            const word = words[i];
            if (!blank) {
                // if previous word is on the same line add separator
                formatted += ' ';
                position++;
            }

            if (position + word.length + (i == words.length - 1 ? 0 : " -".length) <= MAX_LINE_LENGTH) {
                // if there is enough room for the word (including line breaking if necessary)
                formatted += word;
                position += word.length;
                blank = false;
                continue;
            }

            if (!blank) {
                // line is not blank, but the word has not enough room
                if (position <= MAX_LINE_LENGTH - 1) {
                    // on line is room for line breaking, add it and continue on the next line
                    formatted += '-\n' + ''.padEnd(this.nextIdent, ' ');
                    position = this.nextIdent;
                }
                blank = true;
            }

            // try to write the whole long word into lines. The first line could be shorter
            [position, formatted, blank] = JclWriter.formatWord(position, word, i, words, formatted, blank);
        }

        return formatted;
    }

    private static formatWord(position: number, word: string, i: number, words: string[], formatted: string, blank: boolean) {
        let j = MAX_LINE_LENGTH - position - 1;
        while (word.length > 0) {
            if ((i < words.length - 1) && (position + word.length > MAX_LINE_LENGTH - 2) && (word.length < MAX_LINE_LENGTH)) {
                // if the latest piece of word would fulfill the line and it cannot be split, rather split the piece to two lines
                j = word.length - 1;
            }

            if (j >= word.length) {
                // the last piece of word could be written in the line
                formatted += word;
                position = word.length;
                blank = false;
                break;
            } else {
                // there is no enough room for the whole word, put first part and use JCL breaking character
                formatted += word.substring(0, j) + '-\n';
                position = 0;
                word = word.substring(j, word.length);
            }

            j = MAX_LINE_LENGTH - 1;
        }

        return [position, formatted, blank] as const;
    }

    add(command: string) {
        command = command.replaceAll("\r", " ").replaceAll("\n", " ");

        const words = this.parse(command);
        const formatted = this.format(words);

        if (this.commands) {
            this.commands += '\n';
        }
        this.commands += formatted;
    }

    getText(): string {
        return this.commands;
    }

}
