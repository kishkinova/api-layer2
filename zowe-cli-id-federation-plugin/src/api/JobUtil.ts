/**
 * This program and the accompanying materials are made available and may be used, at your option, under either:
 * * Eclipse Public License v2.0, available at https://www.eclipse.org/legal/epl-v20.html, OR
 * * Apache License, version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Copyright Contributors to the Zowe Project.
 */

import {ProfileInfo} from "@zowe/imperative";

export async function getAccount(): Promise<string> {
    let profInfo;
    try {
        profInfo = new ProfileInfo('zowe', {overrideWithEnv: true});
        await profInfo.readProfilesFromDisk();
    } catch (error) {
        return "account";
    }

    const tsoProfAttrs = profInfo.getDefaultProfile("tso");
    if (!tsoProfAttrs) return "account";

    const tsoMergedArgs = profInfo.mergeArgsForProfile(tsoProfAttrs);
    const accountNumberFromProfile = tsoMergedArgs.knownArgs.find(
        arg => arg.argName === "account").argValue as string;

    return accountNumberFromProfile ?? "account";
}
