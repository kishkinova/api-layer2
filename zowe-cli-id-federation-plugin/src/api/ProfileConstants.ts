/**
 * This program and the accompanying materials are made available and may be used, at your option, under either:
 * * Eclipse Public License v2.0, available at https://www.eclipse.org/legal/epl-v20.html, OR
 * * Apache License, version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Copyright Contributors to the Zowe Project.
 */

import {ICommandOptionDefinition} from "@zowe/imperative";
import {Constants} from "./Constants";

export class ProfileConstants {
    public static IDF_CONNECTION_OPTION_GROUP = "IDF Connection Options";

    public static IDF_OPTION_ESM: ICommandOptionDefinition = {
        name: "esm",
        aliases: ["e"],
        description: "The ESM product on the target system",
        required: true,
        type: "string",
        allowableValues: {values: ["RACF", "TSS", "ACF2"]},
        group: ProfileConstants.IDF_CONNECTION_OPTION_GROUP
    };

    public static IDF_OPTION_SYSTEM: ICommandOptionDefinition = {
        name: "system",
        aliases: ["s"],
        description: "The target JES system on which the command will be executed",
        stringLengthRange: [1, Constants.MAX_LENGTH_SYSTEM],
        type: "string",
        group: ProfileConstants.IDF_CONNECTION_OPTION_GROUP
    };

    public static IDF_OPTION_REGISTRY: ICommandOptionDefinition = {
        name: "registry",
        aliases: ["r"],
        description: "The distributed identities registry (e.g., ldaps://enterprise.com, ldap://12.34.56.78:389)",
        required: true,
        type: "string",
        stringLengthRange: [1, Constants.MAX_LENGTH_REGISTRY],
        group: ProfileConstants.IDF_CONNECTION_OPTION_GROUP
    };

    public static IDF_CONNECTION_OPTIONS: ICommandOptionDefinition[] = [
        ProfileConstants.IDF_OPTION_ESM,
        ProfileConstants.IDF_OPTION_SYSTEM,
        ProfileConstants.IDF_OPTION_REGISTRY
    ];
}
