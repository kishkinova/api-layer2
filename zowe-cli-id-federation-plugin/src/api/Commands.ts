/**
 * This program and the accompanying materials are made available and may be used, at your option, under either:
 * * Eclipse Public License v2.0, available at https://www.eclipse.org/legal/epl-v20.html, OR
 * * Apache License, version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Copyright Contributors to the Zowe Project.
 */

import {IHandlerResponseApi, ImperativeError, TextUtils} from "@zowe/imperative";
import { warn } from "console";
import {IIdentity} from "./CsvParser";
import {Constants} from "./Constants";
import {hasValidLength} from "./ValidateUtil";

export class Commands {

    readonly MAX_LENGTH_MAINFRAME_ID = 8;
    readonly MAX_LENGTH_DISTRIBUTED_ID = 246;
    readonly MAX_LENGTH_LABEL = 32;

    constructor(
        private registry: string,
        private identities: IIdentity[],
        private commandTemplate: string,
        private refreshCommand: string,
        private response: IHandlerResponseApi
    ) {
    }

    getCommands(): string[] {
        const commands = this.identities
            .map(identity => this.getCommand(identity))
            .filter(command => command);

        if (!commands.some(Boolean)) {
            this.response.data.setExitCode(Constants.FATAL_CODE);
            throw new ImperativeError({msg: "Error when trying to create the identity mapping."});
        }
        if(this.refreshCommand != null){
            commands.push("");
            commands.push(this.refreshCommand);
        }
        return commands;
    }

    private getCommand(identity: IIdentity): string {
        if(!hasValidLength(identity.mainframeId, this.MAX_LENGTH_MAINFRAME_ID)) {
            warn(`The mainframe user ID '${identity.mainframeId}' has exceeded maximum length of ${this.MAX_LENGTH_MAINFRAME_ID} characters. ` +
           `Identity mapping for the user '${identity.userName}' has not been created.`);
            this.response.data.setExitCode(Constants.WARN_CODE);
            return '';
        }

        if(!hasValidLength(identity.distributedId, this.MAX_LENGTH_DISTRIBUTED_ID)) {
            warn(`The distributed user ID '${identity.distributedId}' has exceeded maximum length of ${this.MAX_LENGTH_DISTRIBUTED_ID} characters. ` +
                `Identity mapping for the user '${identity.userName}' has not been created.`);
            this.response.data.setExitCode(Constants.WARN_CODE);
            return '';
        }

        if(!hasValidLength(identity.userName, this.MAX_LENGTH_LABEL)) {
            warn(`The user name '${identity.userName}' has exceeded maximum length of ${this.MAX_LENGTH_LABEL} characters. ` +
                `Identity mapping for the user '${identity.userName}' has not been created.`);
            this.response.data.setExitCode(Constants.WARN_CODE);
            return '';
        }

        return TextUtils.renderWithMustache(this.commandTemplate, {
            mainframe_id: identity.mainframeId,
            distributed_id: identity.distributedId,
            registry: this.registry,
            user_name: identity.userName,
            escape: function() {
                return function(text: string, render: any) {
                    return render(text).replace(/'/g, "''");
                };
            }
        });
    }

}
