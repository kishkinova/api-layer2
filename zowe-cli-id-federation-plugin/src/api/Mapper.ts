/**
 * This program and the accompanying materials are made available and may be used, at your option, under either:
 * * Eclipse Public License v2.0, available at https://www.eclipse.org/legal/epl-v20.html, OR
 * * Apache License, version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Copyright Contributors to the Zowe Project.
 */

import {ImperativeError, TextUtils} from "@zowe/imperative";
import {RacfCommands} from "./RacfCommands";
import { TssCommands } from "./TssCommands";
import {getAccount} from "./JobUtil";
import * as fs from "fs";
import {CsvParser, IIdentity} from "./CsvParser";
import {JclWriter} from "./JclWriter";

export class Mapper {
    constructor(
        public file: string,
        public esm: string,
        public system: string,
        public registry: string) {
    }

    async map(): Promise<string> {
        const identities = new CsvParser(this.file).getIdentities();
        const commands = this.createSafCommands(identities);
        const jcl = await this.createJcl(commands);
        const fileName = `idf_${this.esm}${this.system ? `_${this.system}` : ""}.jcl`;
        fs.writeFileSync(fileName, jcl);

        const systemMsg = this.system ? ` on the system ${this.system}` : "";
        return `'${fileName}' was created. Review and submit this JCL${systemMsg}.`;
    }

    async createJcl(commands: string[]): Promise<string> {
        const jclTemplate = fs.readFileSync('src/api/templates/job.jcl').toString();
        const account = await getAccount();
        const jclWriter = new JclWriter(1, 2);
        commands.forEach(c => jclWriter.add(c));
        return TextUtils.renderWithMustache(jclTemplate, {
            account: account,
            sysaff: this.system ? `\n/*JOBPARM SYSAFF=${this.system.toUpperCase()}` : "",
            system: this.system ? `system ${this.system}` : "a corresponding system",
            commands: jclWriter.getText()
        });
    }

    createSafCommands(identities: IIdentity[]): string[] {
        let commandProcessor;
        switch (this.esm.toLowerCase()) {
            case "racf": {
                commandProcessor = new RacfCommands(this.registry, identities);
                break;
            }
            case "tss": {
                commandProcessor = new TssCommands(this.registry, identities);
                break;
            }
            case "acf2": {
                break; //TODO: Here will be the code which generate ACF2 commands
            }
            default: {
                const msg = `Unsupported ESM "${this.esm}".` +
                    `Id Federation Plugin supports only the following security systems: RACF, TSS, and ACF2.`;
                throw new ImperativeError({msg});
            }
        }

        return commandProcessor ? commandProcessor.getCommands() : [];
    }


}
