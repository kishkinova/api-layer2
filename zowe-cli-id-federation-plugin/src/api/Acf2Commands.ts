/**
 * This program and the accompanying materials are made available and may be used, at your option, under either:
 * * Eclipse Public License v2.0, available at https://www.eclipse.org/legal/epl-v20.html, OR
 * * Apache License, version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Copyright Contributors to the Zowe Project.
 */

import * as fs from "fs";
import {IIdentity} from "./CsvParser";
import {IHandlerResponseApi} from "@zowe/imperative/lib/cmd/src/doc/response/api/handler/IHandlerResponseApi";
import { Commands } from "./Commands";

export class Acf2Commands extends Commands{

    constructor(
        registry: string,
        identities: IIdentity[],
        response: IHandlerResponseApi
    ) {
        const acf2Template = fs.readFileSync(`${__dirname}/templates/acf2.jcl`).toString();
        const acf2RefreshCommand = fs.readFileSync(`${__dirname}/templates/acf2_refresh.jcl`).toString();
        super(registry,identities,acf2Template,acf2RefreshCommand,response);
    }

    getCommands(): string[] {
        const commands = super.getCommands();
        commands.unshift("ACF", "SET PROFILE(USER) DIVISION(IDMAP)");
        commands.push("END");
        return commands;
    }
}
