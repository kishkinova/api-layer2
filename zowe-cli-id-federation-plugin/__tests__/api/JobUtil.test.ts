/**
 * This program and the accompanying materials are made available and may be used, at your option, under either:
 * * Eclipse Public License v2.0, available at https://www.eclipse.org/legal/epl-v20.html, OR
 * * Apache License, version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 *
 * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
 *
 * Copyright Contributors to the Zowe Project.
 */

import {getAccount} from "../../src/api/JobUtil";
import {ProfileInfo} from "@zowe/imperative/lib/config/src/ProfileInfo";
import {IProfAttrs} from "@zowe/imperative/lib/config/src/doc/IProfAttrs";
import {IProfArgValue, ProfLocType} from "@zowe/imperative";
import {IProfMergedArg} from "@zowe/imperative/lib/config/src/doc/IProfMergedArg";
import {expect, jest, describe, it} from '@jest/globals';

describe("JobUtil unit tests", () => {

    const profAttrs: IProfAttrs = {
        profName: "profName1",
        profType: "profType1",
        profLoc: {
            locType: ProfLocType.TEAM_CONFIG,
            osLoc: ["somewhere in the OS 1", "somewhere in the OS 2"],
            jsonLoc: "somewhere in the JSON file"
        },
        isDefaultProfile: true
    };

    it("should return default value when no default profile found", async () => {
        expect(await getAccount()).toBe("account");
    });

    it("should return default value when error is caught", async () => {
        jest.spyOn(ProfileInfo.prototype, 'readProfilesFromDisk').mockRejectedValue("error");

        expect(await getAccount()).toBe("account");
    });

    it("should successfully parse the file", async () => {
        const profMergedArg: IProfMergedArg = {
            knownArgs:
                [
                    {
                        argName: "account",
                        dataType: "string",
                        argValue: "fake",
                        argLoc: { locType: 0, osLoc: ["location"], jsonLoc: "jsonLoc" },
                        secure: false,
                    },
                ],
            missingArgs: []
        };
        jest.spyOn(ProfileInfo.prototype, 'readProfilesFromDisk').mockImplementation(() => Promise.resolve());
        jest.spyOn(ProfileInfo.prototype, 'getDefaultProfile').mockReturnValue(profAttrs);
        jest.spyOn(ProfileInfo.prototype, 'mergeArgsForProfile').mockReturnValue(profMergedArg);

        const result = await getAccount();

        expect(result).toBe("fake");
        expect(result).toMatchSnapshot();
    });

    it("should return account if account number not defined", async () => {
        const profMergedArg: IProfMergedArg = {
            knownArgs:
                [
                    {
                        argName: "account",
                        dataType: "string",
                        argValue: "",
                        argLoc: { locType: 0, osLoc: ["location"], jsonLoc: "jsonLoc" },
                        secure: false,
                    },
                ],
            missingArgs: []
        };
        jest.spyOn(ProfileInfo.prototype, 'readProfilesFromDisk').mockImplementation(() => Promise.resolve());
        jest.spyOn(ProfileInfo.prototype, 'getDefaultProfile').mockReturnValue(profAttrs);
        jest.spyOn(ProfileInfo.prototype, 'mergeArgsForProfile').mockReturnValue(profMergedArg);

        const result = await getAccount();

        expect(result).toBe("account");
        expect(result).toMatchSnapshot();
    });

});
