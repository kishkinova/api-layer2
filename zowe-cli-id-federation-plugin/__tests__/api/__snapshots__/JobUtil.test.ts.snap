// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobUtil should return account if account number not defined 1`] = `"account"`;

exports[`JobUtil should successfully parse the file 1`] = `"fake"`;
